### PR TITLE
review

### PR DESCRIPTION
**Pull request by boronngo**

### CODE REVIEW

File: src/utils.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import os
+import datetime
+import torch
+
+import torch.nn as nn
+
+
+def create_ckpt_dir():
+    now = datetime.datetime.today()
+    ckpt_dir = 'ckpt/{}{}_{}{}_{}'.format(now.month, now.day,
```
Comment (boronngo) [inline src/utils.py:10]: こここうかなあ。
```python
ckpt_dir = "ckpt/{0:%m%d_%H%M_%S}".format(now)
```
`'ckpt/{}{}_{}{}_{}'`だと後ろまで見ないと何の値が入ってるかわからないし。
%mとか使えば、他の言語とかでも共通だからここには時刻入るんだなって文字列だけでわかるかな。

File: src/utils.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+class Config(dict):
+    def __init__(self, config):
+        self._conf = config
+ 
+    def __getattr__(self, name):
+        if self._conf.get(name) is not None:
+            return self._conf[name]
+
+        return None
```
Comment (boronngo) [inline src/utils.py:35]: これ逆のほうがわかりやすい気がします。
早期returnする場合は意図しない場合を先に除外して、一番下が正常な場合にしたいです。
```python
if self._conf.get(name) is None:
    return None

return self._conf[name]
```

File: src/utils.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        if isinstance(val, dict):
+            dind_keys.append(key)
+        elif not type(val) in [float, int, bool, str]:
+            rm_keys.pop(key)
+
+    for target in dind_keys:
+        val = config.pop(target)
+        config.update(val)
+    for target in rm_keys:
+        del config[targe]
```
Comment (boronngo) [inline src/utils.py:51]: typoです

File: src/train.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        self.dataloader_train = DataLoader(dataset_train,
+                                           batch_size=config.batch_size,
+                                           shuffle=True)
+        self.dataset_val = dataset_val
+        self.criterion = criterion
+        self.optimizer = optimizer
+        self.evaluate = evaluate
+        self.experiment = experiment
+        
+    def iterate(self, num_iter):
```
Comment (boronngo) [inline src/train.py:25]: `num_iter`は使われていないっぽいです

File: src/model.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+        super().__init__()
+        # Define the partial conv layer
+        if sample == 'down-7':
+            self.conv = PartialConv2d(in_ch, out_ch, 7, 2, 3, bias=conv_bias)
+        elif sample == 'down-5':
+            self.conv = PartialConv2d(in_ch, out_ch, 5, 2, 2, bias=conv_bias)
+        elif sample == 'down-3':
+            self.conv = PartialConv2d(in_ch, out_ch, 3, 2, 1, bias=conv_bias)
+        else:
+            self.conv = PartialConv2d(in_ch, out_ch, 3, 1, 1, bias=conv_bias)
```
Comment (boronngo) [inline src/model.py:84]: Pythonってswicth分出来ないんですね…
```python
if sample == 'down-7':
    params = { 'kernel_size': 7, 'stride': 2, 'padding': 3}
...
self.conv = PartialConv2d(in_ch, out_ch, params['kernel_size'], params['stride'], params['padding'], bias=conv_bias)
```
こんな感じですかねえ。個人的にはまとめれる部分はまとめたいなと思います。
ifによってどこが変わってるかわかりやすいと思うので。
（実際に実行していないのでコードは適当です）

File: src/model.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        if self.freeze_enc_bn:
+            for name, module in self.named_modules():
+                if isinstance(module, nn.BatchNorm2d) and 'enc' in name:
+                    module.eval()
+
+
+if __name__ == '__main__':
+    from utils import init_xavier
+    size = (1, 3, 512, 512)
+    input = torch.ones(size)
```
Comment (boronngo) [inline src/model.py:167]: `input`とかはpythonであらかじめ定義されているものなので、上書かれちゃいそうです。  
この辺はpylintとか入れればわかりやすいかと。

File: src/model.py [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+        return feature, mask
+
+    def train(self, mode=True):
+        """
+        Override the default train() to freeze the BN parameters
+        In initial training, BN set to True
+        In fine-tuning stage, BN set to False
+        """
+        super().train(mode)
+        if self.freeze_enc_bn:
```
Comment (boronngo) [inline src/model.py:158]: ここ結構ネストしてるから早期リターンしたほうが見やすいですかね。でも否定だから頭の中的にわかりにくいかなあ。
```python
if not self.freeze_enc_bn:
    return
for name, module in self.named_modules():
    if isinstance(module, nn.BatchNorm2d) and 'enc' in name:
        module.eval()
```

File: src/loss.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    dilated_holes=dialation_holes(hole_mask)
+    colomns_in_Pset=dilated_holes[:, :, :, 1:] * dilated_holes[:, :, :, :-1]
+    rows_in_Pset=dilated_holes[:, :, 1:, :] * dilated_holes[:, :, :-1:, :]
+    loss = torch.sum(torch.abs(colomns_in_Pset*(image[:, :, :, 1:] - image[:, :, :, :-1]))) + \
+        torch.sum(torch.abs(rows_in_Pset*(image[:, :, :1 :] - image[:, :, -1:, :])))
+    return loss
+
+
+if __name__ == '__main__':
+    from config import get_config
```
Comment (boronngo) [inline src/loss.py:128]: 先頭でimportしない意味ってあるんですかね？あんまり良くわかってないです。

Comment (boronngo) [reply src/loss.py:128]: importのスコープが絞れたりするのかしら

File: src/default_config.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
+
+def get_config():
```
Comment (boronngo) [inline src/default_config.py:2]: pythonなら`ConfigParser`とか使ったほうがいいのかな。使ったこと無いから使いにくいのかもしれないけど。  
.pyに設定書かれていると、開かないと設定ファイルであると認識されなさそう。

File: main.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+from src.config import get_config
+from src.dataset import Places2
+from src.model import PConvUNet
+from src.loss import InpaintingLoss, VGG16FeatureExtractor
+from src.train import Trainer
+from src.utils import Config, load_ckpt, create_ckpt_dir
+
+
+# set the config
+config_dict = get_config()
```
Comment (boronngo) [inline main.py:16]: 少なくともfunctionで分けたほうが読みやすいですかねえ。
```python
def config():
    // configする
def define():
    // defineする
def train():
    // trainする
def main():
    // こっから呼ばれる
    config()
    define()
    train()
```

Comment (boronngo) [reply main.py:16]: というか書いてあるコメントでだいたい分けれそう。

**Comment by boronngo** [comment]
大体見ました。見て思ったけどpythonはバッチぐらいでしか書かないからの綺麗な書き方わからない…  
でも、基本的にはちゃんとかけてるのではと思います。  
lintかかってなくて見辛い部分が多いので、入れるだけでも見た目は違うかなあと思います。
